api/models/datastore.go

```diff
@@ -5,21 +5,66 @@ import (
 	"errors"
 )
 
+//TODO path name validation: implementations currently assume/require/rely on:
+// - leading slash
+// - overall legal/scrubbed path (no relative paths, consecutive slashes, etc)
+// - trailing slash allowed and distinct from none
 type Datastore interface {
-	// GetApp returns the app called appName or nil if it doesn't exist
+
+	// GetApp gets an App by name.
+	// Returns ErrDatastoreEmptyAppName for empty appName.
+	// Returns ErrAppsNotFound if no app is found.
 	GetApp(ctx context.Context, appName string) (*App, error)
+
+	// GetApps gets a slice of Apps, optionally filtered by name.
+	// Missing filter or empty name will match all Apps.
 	GetApps(ctx context.Context, filter *AppFilter) ([]*App, error)
+
+	// InsertApp inserts an App. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsAlreadyExists if an App by the same name already exists.
 	InsertApp(ctx context.Context, app *App) (*App, error)
+
+	// UpdateApp updates an App's Config. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsNotFound if an App is not found.
+	//TODO are app.Config's fields added to the existing field or do they replace them?
 	UpdateApp(ctx context.Context, app *App) (*App, error)
```

Comment:
Looking at this again, it seems like ```Config``` should be completely replaced. Otherwise, in order to delete fields, some sort of empty/null/nil-value-means-delete would have to be devised.

Whatever is decided here also likely applies to ```UpdateRoute``` as well.